README
=pod


=head1 SYNOPSYS

    package MyWebService;

    use Dancer2;
    use Dancer2::Plugin::REST;

    prepare_serializer_for_format;

    get '/user/:id.:format' => sub {
        User->find(params->{id});
    };

    get qr{^/user/(?<id>\d+)\.(?<format>\w+)} => sub {
        User->find(captures->{id});
    };

    # curl http://mywebservice/user/42.json
    { "id": 42, "name": "John Foo", email: "john.foo@example.com"}

    # curl http://mywebservice/user/42.yml
    --
    id: 42
    name: "John Foo"
    email: "john.foo@example.com"

=head1 DESCRIPTION

This plugin helps you write a RESTful webservice with Dancer2.

=head1 CONFIGURATION

=head2 serializers

The default format serializer hash which maps a given C<:format> to 
a C<Dancer2::Serializer::*> serializer. Unless overriden in the 
configuration, it defaults to:

    serializers:
      json: JSON
      yml:  YAML
      dump: Dumper

=head1 KEYWORDS

=head2 prepare_serializer_for_format

When this pragma is used, a before filter is set by the plugin to automatically
change the serializer when a format is detected in the URI.

That means that each route you define with a B<:format> param or captures token 
will trigger a serializer definition, if the format is known.

This lets you define all the REST actions you like as regular Dancer2 route
handlers, without explicitly handling the outgoing data format.

Regexp routes will use the file-extension from captures->{'format'} to determine
the serialization format.

=head2 resource

This keyword lets you declare a resource your application will handle.

    resource user =>
        get    => sub { # return user where id = params->{id}   },
        create => sub { # create a new user with params->{user} },
        delete => sub { # delete user where id = params->{id}   },
        update => sub { # update user with params->{user}       };

    # this defines the following routes:
    # GET /user/:id
    # GET /user/:id.:format
    # POST /user
    # POST /user.:format
    # DELETE /user/:id
    # DELETE /user/:id.:format
    # PUT /user/:id
    # PUT /user/:id.:format

=head2 helpers

Some helpers are available. This helper will set an appropriate HTTP status for you.

=head3 status_ok

    status_ok({users => {...}});

Set the HTTP status to 200

=head3 status_created

    status_created({users => {...}});

Set the HTTP status to 201

=head3 status_accepted

    status_accepted({users => {...}});

Set the HTTP status to 202

=head3 status_bad_request

    status_bad_request("user foo can't be found");

Set the HTTP status to 400. This function as for argument a scalar that will be used under the key B<error>.

=head3 status_not_found

    status_not_found("users doesn't exists");

Set the HTTP status to 404. This function as for argument a scalar that will be used under the key B<error>.

=head1 LICENCE

This module is released under the same terms as Perl itself.

=head1 AUTHORS

This module has been written by Alexis Sukrieh C<< <sukria@sukria.net> >> and Franck
Cuny.

=head1 SEE ALSO

L<Dancer2> L<http://en.wikipedia.org/wiki/Representational_State_Transfer>



Changes
Revision history for Dancer2-Plugin-REST

{{$NEXT}}
  [ API CHANGES ]

  [ BUG FIXES ]

  [ DOCUMENTATION ]

  [ ENHANCEMENTS ]

  [ NEW FEATURES ]

  [ STATISTICS ]

0.27 2015-12-29
  - Make changes to the module so that it's compatible with the upcoming
    backward-compatible release. (GH#12, Peter Mottram)

  [ STATISTICS ]
    - code churn: 4 files changed, 75 insertions(+), 67 deletions(-)

0.26 2015-10-15
  [ BUG FIXES ]
    - Set serializer to 'undef' if no format is provided.  (GH#9, Marco
      Masetti, Emil Nicolaie Perhinschi )

  [ STATISTICS ]
    - code churn: 5 files changed, 154 insertions(+), 15 deletions(-)

0.25 2014-11-01
  [ BUG FIXES ]
    - Change tests to use Pack::Test instead of the deprecated
      Dancer2::Test. (GH#3, Davs)

  [ STATISTICS ]
    - code churn: 4 files changed, 62 insertions(+), 41 deletions(-)

0.24 2014-09-01
  [ MISC ]
    - D2 v1.150000 is out, turn v0.23_00 into general availability release.

  [ STATISTICS ]
    - code churn: 3 files changed, 9 insertions(+), 3 deletions(-)

0.23_00 2014-08-13
  [ MISC ]
    - Compatibility with Dancer2's v0.149000_01 changes.

  [ STATISTICS ]
    - code churn: 2 files changed, 61 insertions(+), 62 deletions(-)

0.22 2014-06-05
  [ BUG FIXES ]
    - Fix tests (and code itself) to work with Dancer2 v0.140001. (GH#2,
      veryrusty, Yanick Champoux)

  [ STATISTICS ]
    - code churn: 12 files changed, 272 insertions(+), 302 deletions(-)

0.21 2013-12-07
  [ BUG FIXES ]
    - Auto-serialization was done on the *next* request.

0.20 2013-10-03
  [ BUG FIXES ]
    - Make the plugin Dancer2 v0.10 compatible. (GH#4, jbolila)

0.12 2013-07-06
  [ BUG FIXES ]
    - Still a few issues with the versions in test.

0.11 2013-07-03
  [ BUG FIXES ]
    - Test was checking for wrong version of Dancer2.

0.10 2013-07-03
  - Support routes that use named-captures.
  - Documented serializers configuration.

0.09 2013-02-25
  - Fix to work with Moo v1.000007

0.08 2013-02-24
  - Fork for Dancer2

0.07 2011-11-01
  - Minor updates to fix warnings with Dancer 1.3080
  - Fix POD
  - (Alexis Sukrieh)

0.06 2011-05-28
  [ TEST SUITE ]
    - Skip test t/03_resource.t if Dancer is not >= 1.3059_01 A bug in
      Dancer::Test in previous versions prevent the test from passing.

0.05 2011-03-04
  [ Alexis Sukrieh ]
    - Add all HTTP response status
    - Version 0.03

  [ BUG FIXES ]
    - Fix the test suite with dancer_response
    - Fix dependencies: Dancer 1.3014
    - Version 0.04

  [ Franck Cuny ]
    - New keywords for ReST response status  (status_ok, status_created,
      ...)
    - Version 0.02 * 100% coverage * fix a bug in 'resource' (order of
      route hander definition)
    - Version 0.0001_01 * Support for prepare_serializer_for_format keyword
      automatically set a serializer according to params->{format} *
      Support for resource definition When a resource is defined a set of
      routes are defined in a RESTful manner.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.040.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A plugin for writing RESTful apps with Dancer2",
  "AUTHOR" => "Dancer Core Developers",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dancer2-Plugin-REST",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dancer2::Plugin::REST",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Load" => 0,
    "Dancer2" => "0.149000_01",
    "Dancer2::Plugin" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer2" => "0.149000_01",
    "Dancer2::Core::Request" => 0,
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "HTTP::Request::Common" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "JSON" => 0,
    "Module::Runtime" => 0,
    "Plack::Test" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.27",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Load" => 0,
  "Dancer2" => "0.149000_01",
  "Dancer2::Core::Request" => 0,
  "Dancer2::Plugin" => 0,
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTTP::Request::Common" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON" => 0,
  "Module::Runtime" => 0,
  "Plack::Test" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Dancer2/Plugin/REST.pm
package Dancer2::Plugin::REST;
our $AUTHORITY = 'cpan:SUKRIA';
# ABSTRACT: A plugin for writing RESTful apps with Dancer2
$Dancer2::Plugin::REST::VERSION = '0.27';
use strict;
use warnings;

use Carp 'croak';

use Dancer2::Plugin;
use Class::Load qw/ try_load_class /;

# [todo] - add XML support
my $content_types = {
    json => 'application/json',
    yml  => 'text/x-yaml',
};

register prepare_serializer_for_format => sub {
    my $dsl = shift;

    my $conf        = plugin_setting;
    my $serializers = (
        ($conf && exists $conf->{serializers})
        ? $conf->{serializers}
        : { 'json' => 'JSON',
            'yml'  => 'YAML',
            'dump' => 'Dumper',
        }
    );

    $dsl->hook(
        'before' => sub {
            my $format = $dsl->params->{'format'};
            $format  ||= $dsl->captures->{'format'} if $dsl->captures;

            unless  ( defined $format ) { 
                delete $dsl->app->response->{serializer};
                return;
            }

            my $serializer = $serializers->{$format};
            
            unless( $serializer ) {
                return $dsl->send_error("unsupported format requested: " . $format, 404);
            }

            $dsl->set(serializer => $serializer);
            $dsl->app->set_response( Dancer2::Core::Response->new(
                %{ $dsl->app->response },
                serializer => $dsl->set('serializer'),
            ) );

            my $ct = $content_types->{$format} || $dsl->setting('content_type');
            $dsl->content_type($ct);
        }
    );
};

register resource => sub {
    my $dsl = shift;

    my ($resource, %triggers) = @_;

    my %actions = (
        get    => 'get',
        update => 'put',
        create => 'post',
        delete => 'delete',
    );

    croak "resource should be given with triggers"
      unless defined $resource
             and grep { $triggers{$_} } keys %actions;

    while( my( $action, $code ) = each %triggers ) {
            $dsl->app->add_route( 
                method => $actions{$action},
                regexp => $_,
                code   => $code,
            ) for map { sprintf $_, '/:id' x ($action ne 'create') }
                        "/${resource}%s.:format", "/${resource}%s";
    }
};

register send_entity => sub {
    my ($dsl, $entity, $http_code) = @_;

    $http_code ||= 200;

    $dsl->status($http_code);
    $entity;
};

my %http_codes = (

    # 1xx
    100 => 'Continue',
    101 => 'Switching Protocols',
    102 => 'Processing',

    # 2xx
    200 => 'OK',
    201 => 'Created',
    202 => 'Accepted',
    203 => 'Non-Authoritative Information',
    204 => 'No Content',
    205 => 'Reset Content',
    206 => 'Partial Content',
    207 => 'Multi-Status',
    210 => 'Content Different',

    # 3xx
    300 => 'Multiple Choices',
    301 => 'Moved Permanently',
    302 => 'Found',
    303 => 'See Other',
    304 => 'Not Modified',
    305 => 'Use Proxy',
    307 => 'Temporary Redirect',
    310 => 'Too many Redirect',

    # 4xx
    400 => 'Bad Request',
    401 => 'Unauthorized',
    402 => 'Payment Required',
    403 => 'Forbidden',
    404 => 'Not Found',
    405 => 'Method Not Allowed',
    406 => 'Not Acceptable',
    407 => 'Proxy Authentication Required',
    408 => 'Request Time-out',
    409 => 'Conflict',
    410 => 'Gone',
    411 => 'Length Required',
    412 => 'Precondition Failed',
    413 => 'Request Entity Too Large',
    414 => 'Request-URI Too Long',
    415 => 'Unsupported Media Type',
    416 => 'Requested range unsatisfiable',
    417 => 'Expectation failed',
    418 => 'Teapot',
    422 => 'Unprocessable entity',
    423 => 'Locked',
    424 => 'Method failure',
    425 => 'Unordered Collection',
    426 => 'Upgrade Required',
    449 => 'Retry With',
    450 => 'Parental Controls',

    # 5xx
    500 => 'Internal Server Error',
    501 => 'Not Implemented',
    502 => 'Bad Gateway',
    503 => 'Service Unavailable',
    504 => 'Gateway Time-out',
    505 => 'HTTP Version not supported',
    507 => 'Insufficient storage',
    509 => 'Bandwidth Limit Exceeded',
);

for my $code (keys %http_codes) {
    my $helper_name = lc($http_codes{$code});
    $helper_name =~ s/[^\w]+/_/gms;
    $helper_name = "status_${helper_name}";

    register $helper_name => sub {
        my $dsl = shift;

        $dsl->send_entity(
            ( $code >= 400 ? {error => $_[0]} : $_[0] ),
            $code
        );
    };
}

register_plugin for_versions => [2];

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Dancer2::Plugin::REST - A plugin for writing RESTful apps with Dancer2

=head1 VERSION

version 0.27

=head1 DESCRIPTION

This plugin helps you write a RESTful webservice with Dancer2.

=head1 SYNOPSYS

    package MyWebService;

    use Dancer2;
    use Dancer2::Plugin::REST;

    prepare_serializer_for_format;

    get '/user/:id.:format' => sub {
        User->find(params->{id});
    };

    get qr{^/user/(?<id>\d+)\.(?<format>\w+)} => sub {
        User->find(captures->{id});
    };

    # curl http://mywebservice/user/42.json
    { "id": 42, "name": "John Foo", email: "john.foo@example.com"}

    # curl http://mywebservice/user/42.yml
    --
    id: 42
    name: "John Foo"
    email: "john.foo@example.com"

=head1 CONFIGURATION

=head2 serializers

The default format serializer hash which maps a given C<:format> to 
a C<Dancer2::Serializer::*> serializer. Unless overriden in the 
configuration, it defaults to:

    serializers:
      json: JSON
      yml:  YAML
      dump: Dumper

=head1 KEYWORDS

=head2 prepare_serializer_for_format

When this pragma is used, a before filter is set by the plugin to automatically
change the serializer when a format is detected in the URI.

That means that each route you define with a B<:format> param or captures token 
will trigger a serializer definition, if the format is known.

This lets you define all the REST actions you like as regular Dancer2 route
handlers, without explicitly handling the outgoing data format.

Regexp routes will use the file-extension from captures->{'format'} to determine
the serialization format.

=head2 resource

This keyword lets you declare a resource your application will handle.

    resource user =>
        get    => sub { # return user where id = params->{id}   },
        create => sub { # create a new user with params->{user} },
        delete => sub { # delete user where id = params->{id}   },
        update => sub { # update user with params->{user}       };

    # this defines the following routes:
    # GET /user/:id
    # GET /user/:id.:format
    # POST /user
    # POST /user.:format
    # DELETE /user/:id
    # DELETE /user/:id.:format
    # PUT /user/:id
    # PUT /user/:id.:format

=head2 helpers

Some helpers are available. This helper will set an appropriate HTTP status for you.

=head3 status_ok

    status_ok({users => {...}});

Set the HTTP status to 200

=head3 status_created

    status_created({users => {...}});

Set the HTTP status to 201

=head3 status_accepted

    status_accepted({users => {...}});

Set the HTTP status to 202

=head3 status_bad_request

    status_bad_request("user foo can't be found");

Set the HTTP status to 400. This function as for argument a scalar that will be used under the key B<error>.

=head3 status_not_found

    status_not_found("users doesn't exists");

Set the HTTP status to 404. This function as for argument a scalar that will be used under the key B<error>.

=head1 LICENCE

This module is released under the same terms as Perl itself.

=head1 AUTHORS

This module has been written by Alexis Sukrieh C<< <sukria@sukria.net> >> and Franck
Cuny.

=head1 SEE ALSO

L<Dancer2> L<http://en.wikipedia.org/wiki/Representational_State_Transfer>

=head1 AUTHOR

Dancer Core Developers

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Alexis Sukrieh.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
